cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -126,42 +127,82 @@ static void show_buffers(const char *line, int offset,
 	fprintf(stderr, ">%s<\n", line);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i) {
+	for (size_t i = 0; i < strlen(line); ++i) {
 		int c = entry->line_buf[i];
 		fputc(c == 0 ? '@' : c == 1 ? '#' : c, stderr);
 	}
 	fputc('<', stderr);
 	fputc('\n', stderr);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i)
+	for (size_t i = 0; i < strlen(line); ++i)
 		fputc('=', stderr);
 	fputc('<', stderr);
 	fputc('\n', stderr);
 #endif				// MOUNTINFO_DEBUG
 }
 
+static bool is_octal_digit(char c)
+{
+	return c >= '0' && c <= '7';
+}
+
 static char *parse_next_string_field(sc_mountinfo_entry * entry,
-				     const char *line, int *offset)
+				     const char *line, size_t * offset)
 {
-	int offset_delta = 0;
-	char *field = &entry->line_buf[0] + *offset;
-	if (line[*offset] == ' ') {
-		// Special case for empty fields which cannot be parsed with %s.
-		*field = '\0';
-		*offset += 1;
-	} else {
-		int nscanned =
-		    sscanf(line + *offset, "%s%n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		*offset += offset_delta;
-		if (line[*offset] == ' ') {
-			*offset += 1;
+	const char *input = &line[*offset];
+	char *output = &entry->line_buf[*offset];
+	const char *s;		// Used for escape logic.
+	size_t input_idx = 0;
+	size_t output_idx = 0;
+
+	// Scan characters until we run out of memory to scan or we find a space.
+	// The kernel uses simple octal escape sequences for the following:
+	// space, tab, newline, backwards slash. Everything else is copied
+	// verbatim.
+	bool stop = false;
+	for (int c; c = input[input_idx], c != '\0' && !stop;
+	     input_idx++, output_idx++) {
+		switch (c) {
+		case ' ':
+			// Fields are space delimited. Represent the space
+			// as the end-of-string marker, skip over it, and stop parsing.
+			output[output_idx] = '\0';
```

Comment:
Is this redundant? We do the same when we leave the loop? (well, not exactly the same, we do `output[output_idx] = 0;` there). I think if its redundant we should remove it. If not we should explain why not and also use the same style "0" or '\0'.

Reply:
I think this is not redundant because `output_idx` is incremented in the for loop at the end of this iteration. The one at the bottom of the loop is used when the input string terminates (note that we don't copy the terminating '\0' in the loop itself.

I can tweak the code at the bottom of the loop to use `'\0'` rather than `0` and add a comment to explain this.

Reply:
Indeed, you are right. Thats fine then, I think the '\0' vs 0 is not super important, I just noticed that we are a bit inconsistent (alt in the loop check we use '\0'). But it makes no difference in practise so *meh*

Reply:
I changed it to be consistent.